ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,297 @@
+stored in a tuple which is hashable in Python.
+"""
+SpanAggrKey = typing.Tuple[
+    str,  # name
+    str,  # service
+    str,  # resource
+    str,  # type
+    int,  # http status code
+    bool,  # synthetics request
+]
+
+
+class SpanAggrStats(object):
+    """Aggregated span statistics."""
+
+    __slots__ = ("hits", "top_level_hits", "errors", "duration", "ok_distribution", "err_distribution")
+
+    def __init__(self):
+        self.hits = 0
+        self.top_level_hits = 0
+        self.errors = 0
+        self.duration = 0
+        # Match the relative accuracy of the sketch implementation used in the backend
+        # which is 0.775%.
+        self.ok_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+        self.err_distribution = LogCollapsingLowestDenseDDSketch(0.00775, bin_limit=2048)
+
+
+def _span_aggr_key(span):
+    # type: (Span) -> SpanAggrKey
+    """Return a hashable key that can be used to aggregate similar spans."""
+    service = span.service or ""
+    resource = span.resource or ""
+    _type = span.span_type or ""
+    status_code = span.get_tag("http.status_code") or 0
+    synthetics = span.context.dd_origin == "synthetics"
+    return span.name, service, resource, _type, int(status_code), synthetics
+
+
+class SpanStatsProcessorV06(PeriodicService, SpanProcessor):
+    """SpanProcessor for computing, collecting and submitting span metrics to the Datadog Agent."""
+
+    def __init__(self, agent_url, interval=None, timeout=1.0, retry_attempts=3, reuse_connections=None):
+        # type: (str, Optional[float], float, int, Optional[bool]) -> None
+        if interval is None:
+            interval = float(os.getenv("_DD_TRACE_STATS_WRITER_INTERVAL") or 10.0)
+        if reuse_connections is None:
+            reuse_connections = asbool(os.getenv("_DD_TRACE_STATS_WRITER_REUSE_CONNECTIONS", False))
+        super(SpanStatsProcessorV06, self).__init__(interval=interval)
+        self.start()
+        self._agent_url = agent_url
+        self._timeout = timeout
+        # Have the bucket size match the interval in which flushes occur.
+        self._bucket_size_ns = int(interval * 1e9)  # type: int
+        self._buckets = defaultdict(
+            lambda: defaultdict(SpanAggrStats)
+        )  # type: DefaultDict[int, DefaultDict[SpanAggrKey, SpanAggrStats]]
+        self._endpoint = "/v0.6/stats"
+        self._connection = None  # type: Optional[ConnectionType]
+        self._headers = {
+            "Datadog-Meta-Lang": "python",
+            "Datadog-Meta-Tracer-Version": ddtrace.__version__,
+            "Content-Type": "application/msgpack",
+        }  # type: Dict[str, str]
+        self._lock = Lock()
+        self._enabled = True
+        self._reuse_connections = reuse_connections
+        self._retry_request = tenacity.Retrying(
+            wait=tenacity.wait_random_exponential(
+                multiplier=0.618 * self.interval / (1.618 ** retry_attempts) / 2, exp_base=1.618
```

Comment:
are these copy/pasted magic numbers? comments? constants?

Reply:
The golden ratio is pretty magical

but yes these are copy pasted from the trace writer, not sure what the rationale was behind them

Reply:
https://github.com/DataDog/dd-trace-py/blob/822515d370d617e14cba8f5356388de04f9b2201/ddtrace/internal/writer.py#L311-L312